File: learn.py
from flask import Flask, jsonify
import json
import requests

app = Flask(__name__)

OLLAMA_URL = "http://localhost:11434/api/generate"



# Function to generate coding questions using CodeLlama
def generate_questions():
    final_prompt = """Generate a set of 10 coding questions which consist of Java Beginner and Intermediate level coding questions based on:
    - Output
    - Variables
    - Data types
    - Typecasting
    - Operators
    - Strings
    - Booleans
    - If-else
    - Switch

    Each question should include:
    - **Question**
    - **Solution**
    - **Hint/Steps to solve**.

    Do not repeat any topic or question.
    
    Format the response strictly in JSON format.
    """

    data = {
        "model": "LearnCode",  # Your specific CodeLlama model name
        "prompt": final_prompt,
        "stream": False
    }

    try:
        response = requests.post(OLLAMA_URL, json=data, headers={"Content-Type": "application/json"})

        if response.status_code == 200:
            response_json = response.json()  # Directly parse JSON
            actual_response = response_json.get("response", "")

            try:
                questions_list = json.loads(actual_response)  # Ensure it's valid JSON
                return jsonify(questions_list)
            except json.JSONDecodeError:
                return jsonify({"error": "Failed to parse JSON response"})

        else:
            return jsonify({"error": f"API error: {response.text}"})

    except requests.RequestException as e:
        return jsonify({"error": f"Request failed: {str(e)}"})

# API endpoint to get generated questions
@app.route("/generate-questions", methods=["GET"])
def get_questions():
    return generate_questions()

if __name__ == "__main__":
    app.run(debug=True)

File: server/learn.js
const express = require("express");
const { GoogleGenerativeAI } = require("@google/generative-ai");
require("dotenv").config(); // Load API key from .env file

const app = express();
const PORT = 5002;
const GEMINI_API_KEY = "AIzaSyCXS8cqy_sJSRRjAh5rW9Q1sToqigK_5Nw"; // Store API key in .env file

// Initialize Gemini AI
const genAI = new GoogleGenerativeAI("AIzaSyCXS8cqy_sJSRRjAh5rW9Q1sToqigK_5Nw");

const generateQuestions = async () => {
  const finalPrompt = `
  You are a Java coding problem generator. Your task is to create 10 Java-based coding questions of varying difficulty from beginner, intermediate. Follow these guidelines:
- Generate a mix of problems: 5 beginner, 5 intermediate.
- Ensure questions cover different concepts such as output, variables, datatypes, typecasting, strings, booleans, if-else, switch, loops.

Generate questions, solution and hints.

- Return all 10 questions, solutions and hints in a JSON format.
{
    Question:""
    Solution:""
    Hint:""
}
    

  `;

  try {
    const model = genAI.getGenerativeModel({ model: "gemini-2.0-flash" }); // Use latest stable version
    const result = await model.generateContent(finalPrompt);
    const response = await result.response.text(); // Extract text response

    // Gemini sometimes wraps JSON in markdown (```json ... ```)
    const cleanedResponse = response.replace(/```json|```/g, "").trim();

    return JSON.parse(cleanedResponse); // Parse JSON correctly
  } catch (error) {
    console.error("Error generating questions:", error);
    return { error: `Request failed: ${error.message}` };
  }
};

// API endpoint to get generated questions
app.get("/generate-questions", async (req, res) => {
  const questions = await generateQuestions();
  res.json(questions);
});

app.listen(PORT, () => {
  console.log(`Server running on http://localhost:${PORT}`);
});
